sources/updater/updog/src/transport.rs

```diff
@@ -1,61 +1,78 @@
     pub inner: HttpTransport,
-    parameters: RefCell<Vec<(String, String)>>,
+    parameters: QueryParams,
 }
 
-impl HttpQueryTransport {
-    pub fn new() -> Self {
-        Self {
-            inner: HttpTransport::new(),
-            parameters: RefCell::new(vec![]),
+impl QueryParams {
+    pub(crate) fn add_params_to_url(&self, mut url: Url) -> Url {
+        let mut params = match self.0.write() {
+            Err(e) => {
+                // a thread died while holding a lock to the params. unlikely to occur.
+                error!("unable to add query params to HTTP call: {}", e);
+                return url;
+            }
+            Ok(lock_result) => lock_result,
+        };
+        params.sort_by(|(a, _), (b, _)| a.cmp(b));
+        for (key, val) in params.iter() {
+            url.query_pairs_mut().append_pair(&key, &val);
         }
+        url
     }
 
-    /// Try to borrow a mutable reference to parameters; returns an error if
-    /// a borrow is already active
-    pub fn queries_get_mut(
-        &self,
-    ) -> Result<std::cell::RefMut<'_, Vec<(String, String)>>, BorrowMutError> {
-        self.parameters.try_borrow_mut()
-    }
-
-    /// Set the query string appended to tough requests, sorting the queries
-    /// by key name first
-    fn set_query_string(&self, mut url: Url) -> Url {
-        if let Ok(mut queries) = self.parameters.try_borrow_mut() {
-            queries.sort_by(|(a,_), (b,_)| a.cmp(b));
-
-            for (key, val) in queries.iter() {
-                url.query_pairs_mut().append_pair(&key, &val);
+    pub(crate) fn add<S1, S2>(&self, key: S1, val: S2) -> ()
+    where
+        S1: AsRef<str>,
+        S2: AsRef<str>,
+    {
+        let mut params = match self.0.write() {
+            Err(e) => {
+                // a thread died while holding a lock to the params. unlikely to occur.
+                error!(
+                    "unable to add query param '{}={}': {}",
+                    key.as_ref(),
+                    val.as_ref(),
+                    e
+                );
+                return;
             }
-        } else {
-            // We can't sort the actual data at the moment, but we can sort
-            // what we append to the URL.
-            let mut queries = self.parameters.borrow().clone();
-            queries.sort_by(|(a,_), (b,_)| a.cmp(b));
+            Ok(lock_result) => lock_result,
+        };
+        params.push((key.as_ref().into(), val.as_ref().into()))
+    }
+}
 
-            for (key, val) in queries {
-                url.query_pairs_mut().append_pair(&key, &val);
-            }
+impl HttpQueryTransport {
+    pub fn new() -> Self {
+        Self {
+            inner: HttpTransport::new(),
+            parameters: QueryParams::default(),
         }
+    }
 
-        url
+    /// Obtain a shared pointer to the query params for this transport.
+    pub fn query_params(&self) -> QueryParams {
+        self.parameters.clone()
```

Comment:
```suggestion
        Arc::clone(&self.parameters)
```
It's a little more idiomatic and makes it clear that we're just making another handle, not a data clone.